### ISSUE TITLE
wasm implementation of descendantsOfType causing out of bounds error

### ISSUE THREAD

**Issue by Razzeee** [tree-sitter/tree-sitter#394]
Hey,

it seems like the new decendantsOfType implementation has some problems, I'm getting this from a file.

```
RuntimeError: memory access out of bounds
    at wasm-function[96]:210
    at wasm-function[667]:962
    at Object.Module._ts_node_descendants_of_type_wasm (/home/razze/Development/elm-language-client-vscode/server/node_modules/web-tree-sitter/tree-sitter.js:1:33824)
    at Node.descendantsOfType (/home/razze/Development/elm-language-client-vscode/server/node_modules/web-tree-sitter/tree-sitter.js:1:49842)
    at Function.findAllFunctionDeclarations (/home/razze/Development/elm-language-client-vscode/server/out/util/treeUtils.js:303:30)
    at Function.findFunction (/home/razze/Development/elm-language-client-vscode/server/out/util/treeUtils.js:254:32)
    at Function.getModuleNameAndExposing (/home/razze/Development/elm-language-client-vscode/server/out/util/treeUtils.js:93:51)
    at Forest.setTree (/home/razze/Development/elm-language-client-vscode/server/out/forest.js:55:52)
    at Server.initialize (/home/razze/Development/elm-language-client-vscode/server/out/server.js:125:24)
    at Server.registerProviders (/home/razze/Development/elm-language-client-vscode/server/out/server.js:161:14)
```

It's happening when I call decendantsOf Type on a node in this file https://github.com/elm/core/blob/1.0.0/src/Array.elm

**Comment by maxbrunsfeld** [comment]
Can you write up a minimal JavaScript file that reproduces the problem when run?

**Comment by Razzeee** [comment]
I'll try

**Comment by Razzeee** [comment]
So here's what else I learned so far.

The error comes up every time I try to load a certain project. https://github.com/Razzeee/htbah-character-creator/tree/master to be exact. I tested this on windows and linux, both get the same error.

I build a small project on linux, that:
1. Loads tree sitter + elm definitions via wasm
2. Loads the file that seems to cause it
3. Run `descendantsOfType` on that tree.

I can't get the error that way, so there has to be something else going on.
I also tried to call `descendantsOfType` on that tree with a while loop, still doesn't seem to cause the error.

**Comment by Razzeee** [comment]
I'm also getting this error in another project on windows (linux not tested). Might be completely unrelated and not even tree-sitter. Bu as far as I'm aware tree sitter is the only native lib I'm using. Unfortunately there is no stack trace available.

```
abort("Assertion failed: self->ref_count > 0, at: lib/src/./stack.c,89,stack_node_retain"). Build with -s ASSERTIONS=1 for more info.
```
So it seems to match this method https://github.com/tree-sitter/tree-sitter/blob/bc5e71bcc2095552d655ad83221bdbf78d799224/lib/src/stack.c#L86

**Comment by Razzeee** [comment]
Was able to replicate it in a SSCCE, still not as small as I would like it. But hopefully small enough.

https://github.com/Razzeee/tree-sitter-bounds-error

**Comment by Razzeee** [comment]
One weird thing I just figured out, when you add a line break after https://github.com/Razzeee/tree-sitter-bounds-error/blob/master/Array.elm#L1004
the error goes away

**Comment by Razzeee** [comment]
Another thing that's happening, when we're able to parse the files the first time around, is that cpu will jump up to 100% and stay there, when we update the tree and run decendantsOfType on it (findFunction is doing that).

![image](https://user-images.githubusercontent.com/1693421/61593172-1664b700-abaa-11e9-8488-5a3a7e3eedf2.png)

**Comment by Razzeee** [comment]
I was able to condense https://github.com/Razzeee/tree-sitter-bounds-error down a bit more.

One thing that's also interesting, is that I now catch the error on `descendantsOfType` and log them. Which means I get examples of this error happening while going through the foreach.

**Comment by maxbrunsfeld** [comment]
Can you write the repro script in plain JavaScript instead of TypeScript? It looks like the TypeScript compiler has translated an `async`/`await` syntax into a very complex and difficult-to-read function that's designed to run on very old JS engines. I don't think this is needed though - JS has natively supported `yield`, `async` and `await` for a long time now.

**Comment by Razzeee** [comment]
That might be due to the target here https://github.com/Razzeee/tree-sitter-bounds-error/blob/master/tsconfig.json#L6 I kept it on the same that vscode seems to use. I can look at doing it in js, but I'm not that fit in that eco system.

On another node, I did a workaround now, that's handling the `descendantsOfType` logic in typescript. Which obviously will be slower and is just a workaround. But I'm not getting the error anymore as far as I can tell. But the 100% cpu issue is still there. Will try to track that one down next.
https://github.com/elm-tooling/elm-language-server/commit/04ea9d7e9d4bcb6556f52607c864110dcdc7f53b#diff-21245dcc42cbe9ffe996a08b05568766R1066
